### ISSUE TITLE
Ability to print chat

### ISSUE THREAD

**Issue by levicki** [oobabooga/text-generation-webui#2745]
**Description**

When you chat with the bot and send pictures, and use Stable Diffusion to generate images in responses you get a nicely ilustrated story. Sadly, if you try to invoke print dialog from the browser, it tries to print the whole UI with the chat part cut off.

It would be nice to have have a Print Chat button which would invoke browser print dialog just for the chat `div` with images stretched to full page width (A4 or Letter), and without surrounding UI. The idea is to actually be able to use Print to PDF to save the whole chat with photos into a PDF document.

I think this could be even hacked together with Tampermonkey script (I did similar thing for one website), but I am not so familiar with Gradio that much so it would take me more time to figure it out than if someone who knows it went in and added a print button.

**Comment by missionfloyd** [comment]
Something like this should work (`chat.css`). It's not quite there though,
```css
@media print {
    body {
        visibility: hidden;
    }

    .chat {
        visibility: visible;
        position: absolute;
        left: 0;
        top: 0;
        max-width: none;
        max-height: none;
        width: 100%
        height: fit-content;
    }
}
```

**Comment by levicki** [comment]
Thanks, I will try that.

**Comment by levicki** [comment]
@missionfloyd I experimented a bit and I could get it to work partially but it cuts the last couple of messages. Also, I couldn't get it to respect margins and colors fully (even with CSS override).

I think that the best solution would be to add a Print Chat button to gradio UI and invoke Javascript on click -- something like this perhaps:
```
const features = 'left=200,top=200,width=1060,height=780,popup';
const chat = document.querySelector('#chat').innerHTML;
const type = document.implementation.createDocumentType('html', '', '');
const doc = document.implementation.createDocument('http://www.w3.org/1999/xhtml', 'html', type);
doc.documentElement.setAttribute('lang', document.documentElement.lang);
const head = doc.createElement('head');
const meta = doc.createElement('meta');
meta.setAttribute('charset', 'UTF-8');
head.appendChild(meta);
const title = doc.createElement('title');
const title_text = doc.createTextNode('Chat Printout');
title.appendChild(title_text);
head.appendChild(title);
const script = document.createElement('script');
const code = `window.onload = function() {
	window.print();
	setTimeout(function() {
		window.close();
	}, 1);
};`;
script.setAttribute('src', URL.createObjectURL(new Blob([code]), { type: 'text/javascript' }));
head.appendChild(script);
doc.documentElement.appendChild(head);
const body = doc.createElement('body');
const div = doc.createElement('div');
div.setAttribute('style', 'font-family: "Segoe UI";'); // Add any other styles as needed
div.setHTML(chat);
body.appendChild(div);
doc.documentElement.appendChild(body);
const html = (new XMLSerializer).serializeToString(doc);
const url = URL.createObjectURL(new Blob([html], { type: "text/html" }));
const win = window.open(url, '', features);
```
If anyone knows a better way do let me know.

**Comment by Woisek** [comment]
> When you chat with the bot and send pictures, and use Stable Diffusion to generate images in responses you get a nicely illustrated story. 

Sorry if I hijack this thread, but ... I haven't found anything about this "send pictures". It sounds interesting, could you maybe shortly explain how it works, or point me to a page where I can read it up?
Thanks! 👍

**Comment by missionfloyd** [comment]
> I haven't found anything about this "send pictures"

It's an [extension](https://github.com/oobabooga/text-generation-webui/blob/main/docs/Extensions.md). Check the `Interface Mode` tab.

**Comment by Woisek** [comment]
@missionfloyd 
Many thanks for the hint. But what does it actually do? How is it used? 🤔
I see all those extension in text generation but no description on what they do. What did I miss here?

**Comment by levicki** [comment]
@Woisek It is considered rude to hijack issues like you just did &mdash; please don't do that in the future because it generates noise for everyone subscribed to the issue. This repository has a Discussions tab where you can go and ask questions for which you can't already find answers in the Wiki tab.

**Comment by Woisek** [comment]
@levicki 
I know what hijacking is considered, That's why I apologized for it. And since I don't get some (satisfying) answers in your mentioned discussions, I hoped for one in this, very rarely done from me, exception. You probably haven't looked at it yet yourself, but the wiki has unfortunately nothing to offer to help new users to understand this whole repo 'oobabooga' and how it has or should be configured. But so that you are not disturbed by "noise", I will delete my postings. And feel free to ignore or block me if I bothers you that much.
Otherwise I'm thankfully for any help I could get, like another here, too, I guess.

**Comment by levicki** [comment]
@Woisek 

> I know what hijacking is considered, That's why I apologized for it. 

You are consciously violating the code of conduct which makes your apology meaningless. This is not a discussion forum &mdash; this is a code repository issue tracker where developers who answered the reported issue get an email for every message you post.

The reason not to post offtopic messages is to not disturb developers because they are doing this work for free and you don't want to discourage them from solving the reported issue by making it hard for them to track the original topic.

> And since I don't get some (satisfying) answers in your mentioned discussions...

You didn't pay for this software, nor you are paying for support.

Therefore, nobody owes you any answers here &mdash; it's all just good will of some good people and if you annoy them by doing stuff you claim you are aware you shouldn't be doing, then you are just reducing your chances to get help.

> You probably haven't looked at it yet yourself...

How do you think I learned about and installed those extensions?

It's all there, you just need to find it.

> but the wiki has unfortunately nothing to offer to help new users to understand this whole repo

First, that's absolutely not true, and claiming that shows how little effort you put into trying to find answers by yourself.

Second, you can't really expect to just download a one click installer and run some AI stuff without reading anything on the subject first to at least familiarize yourself with terminology, environment, tools, and technologies used.

Third, there are numerous tutorials on rentry.org and hours of video on youtube.com explaining all this stuff including what you are asking about.

>  ... if I bothers you that much.

It's not about me, it's about open-source developer community this project belongs to.

They expect people who download this project, which they made available to everyone for free, to do their homework before coming to ask for help.

Not doing so, and then also breaking the code of conduct, and finally when confronted claiming that they didn't provide enough hand-holding to you so you can easily use this rather complex software is a sure-fire way to antagonize them.

TL;DR &mdash; read the [Wiki](https://github.com/oobabooga/text-generation-webui/tree/main/docs#readme). If you can't find answers there, [start a discussion](https://github.com/oobabooga/text-generation-webui/discussions/new?category=q-a) and ask for help on a specific subject. Never post offtopic comments in other people's reported issues.

Hope that helps.

**Comment by oobabooga** [comment]
Please stop the drama, he just asked the question.

**Comment by levicki** [comment]
> Please stop the drama, he just asked a question.

OK.

**Comment by Woisek** [comment]
OK, for the last time off-topic, no more messages after that.

First, I know exactly what i did. I'm not just on the net since yesterday.
Second, I'm not an idiot, I've been using computers for more than 30 years and I've been programming websites for more than 12 years. I think I already have some experience with technology. Furthermore, I'm absolutely independent in learning, I have no problem to look for things if they are findable. and for oobabooge, unfortunately, there is absolutely nothing about what and how you have to set something, what effects it has, etc.

You think the wiki has all the answers? 
Fine, where exactly does it say:
What instruction templates are and which one to use when?
How should User string, Bot string and Context be executed and why?
What is the command for chat-instruct mode and what must/should be entered there?
All the generation parameters presets. When must/should you use which preset?

Should I continue?
If you can show me where I can read all this and more, then keep it coming. And since the oobabooba subreddit is still closed, I don't have the possibility to get information there.

Furthermore, what kind of weird argument and treatment should it be, just because something is free, that one is treated so derogatory from above? Yes, it's nice when someone takes the trouble to offer it to the community. I personally can't do that, but I do my best to at least help out where I can. Many others do not. And when someone like you comes wobbling here and tries to let go of unnecessary teachings instead of helping, then that's not exactly motivating.

Finally, I hereby apologize to @oobabooga  for this unexpected excess that has arisen here, but I could not foresee that.

So, and with that I am now out of this thread.

~eoc~
